### DISCUSSION TITLE
about the class Writer 简化

### DISCUSSION THREAD

**Discussion by nenge123** [gildas-lormeau/zip.js#378] [Ideas]
不如增加一个流ReadableStream?
await entry.getData(new zip.Uint8ArrayWriter(),opt) =>  await entry.getData('json|Uint8Array',opt) ?
Uint8ArrayWriter|BlobWriter 简化~
```javascript
	class Writer extends Stream {

		constructor(returntype) {
			super();
			const writer = this;
			if(returntype)writer.returntype = returntype;
			writer.dataReader = new Response(new ReadableStream({
				start(controller){
					return writer.saveChunk(controller);
				}
			}));
			const writable = new WritableStream({
				write(chunk) {
					writer.enqueue(chunk);
					//return writer.writeUint8Array(chunk);
				},
				close(){
					writer.closeChunk();
				}
			});
			Object.defineProperty(writer, PROPERTY_NAME_WRITABLE, {
				get() {
					return writable;
				}
			});
		}
		enqueue(chunk){
			this.controller.enqueue(chunk);
		}
		saveChunk(controller){
			this.controller = controller;
			return new Promise(complete=>this.complete=complete);
		}
		closeChunk(){
			this.controller.close();
			this.complete();
			//console.log(this.dataReader.arrayBuffer());
			//console.log(this.dataReader.blob());
			console.log(this.dataReader);
		}
		async getData(){
			let type = this.returntype||'Uint8Array';
			if(type=='Uint8Array') return new Uint8Array(await this.dataReader.arrayBuffer())
			return await this.dataReader[type]();
		}
		writeUint8Array() {
			// abstract
		}
	}
```

**Comment by gildas-lormeau** [comment]
Multiple classes inherit from [`Writer`](https://gildas-lormeau.github.io/zip.js/api/classes/Writer.html) (see Hierarchy). For example, your changes introduces some issues in [`Data64URIWriter`](https://gildas-lormeau.github.io/zip.js/api/classes/Data64URIWriter.html) like breaking the constructor code. Also, `writeUint8Array()` does nothing in your implementation and this method is used by classes inheriting from `Writer`. Actually, it's hard for me to understand what is the point of your suggestion in the case of `Data64URIWriter`, for example.

On the other hand, I'm not a fan of APIs relying on strings, e.g. `entry.getData("Uint8Array")`, `entry.getData("json")`, or `entry.getData("blob")` (I guess TypeScript developers agree). The purpose of the `Entry` class (and `ZipReader`, `ZipWriter` classes) is to be as low-level as possible. The high-level API in zip.js is [`fs.FS`](https://gildas-lormeau.github.io/zip.js/api/classes/FS.html). This is where I would place such code. However, I prefer using explicit APIs like `ZipFileEntry#getBlob`, `ZipFileEntry#getData64URI`, `ZipFileEntry#getUint8Array` etc., see https://gildas-lormeau.github.io/zip.js/api/classes/ZipFileEntry.html.

**Comment by nenge123** [comment]
@gildas-lormeau I see the "BlobWriter" ,  
why the "Uint8ArrayWriter " not be like this?
then can remove the ```for()```
```javascript
class Uint8ArrayWriter extends BlobWriter {
  async getData(){
	  const blob = await super.getData();
	  console.log(blob);
	  return new Uint8Array(await blob.arrayBuffer());
  }
}
	class TextWriter extends BlobWriter{
		constructor(encoding){
			super(encoding);
		}
		async getData(){
			const blob = await super.getData();
			return await blob.text();

		}
	}
	class Data64URIWriter extends Uint8ArrayWriter{
		constructor(contentType){
			super();
			this.contentType = contentType;
		}
		async getData(){
			return "data:" + (this.contentType || "") + ";base64,"+btoa(await super.getData());

		}
	}
```

**Reply by gildas-lormeau** [reply]
Because it will consume at least twice as much RAM when calling `blob.arrayBuffer` or `btoa` for a short period of time (i.e. until the GC will destroy unused objects). The purpose of the implementation of `Uint8ArrayWriter` and `Data64URIWriter` in zip.js is to avoid this.

**Reply by gildas-lormeau** [reply]
Note however that nothing prevents you to use your own custom implementation of `Writer` classes. Actually, zip.js was designed for this, see https://github.com/gildas-lormeau/zip.js/blob/master/tests/all/test-custom-io.js for example.
